File: scala/0206-reverse-linked-list.scala
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
    def reverseList(head: ListNode): ListNode = {
        if (head == null || head.next == null) {
            return head
        }
        
        val reversedListHead = reverseList(head.next)
        head.next.next = head
        head.next = null
        
        return reversedListHead
    }
}

File: go/0206-reverse-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	reversedListHead := reverseList(head.Next)
	head.Next.Next = head
	head.Next = nil
	return reversedListHead
}

// Iterative version
// func reverseList(head *ListNode) *ListNode {
//     var prev *ListNode
//     curr := head
//
//     for curr != nil {
//         tmp := curr.Next
//         curr.Next = prev
//         prev = curr
//         curr = tmp
//     }
//
//     return prev
// }